### ISSUE TITLE
7.0: Querying the mapped value ... before task ... has completed is not supported caused by LazyPublishArtifact

### ISSUE THREAD

**Issue by vlsi** [gradle/gradle#16777]
### Expected Behavior

It should be possible to add lazy tasks as an artifact

### Current Behavior

6.8.3 produces a warning, and 7.0 fails the build.

CI failure: https://github.com/vlsi/gradle-issue-reproducers/pull/5/checks?check_run_id=2317062953#step:4:227

`/build.gradle.kts`:

```kotlin
plugins {
    `java-library`
}

val zips by configurations.creating {
    isCanBeResolved = true
    isCanBeConsumed = false
}

dependencies {
    zips(project(":dist", "zips"))
}

val printDistFiles by tasks.registering {
    inputs.files(zips)
    doLast {
        val names = zips.map { it.name }.sorted()
        val expected = listOf("buildZip.zip", "buildZip.zip.sha512")
        require(names == expected) {
            "expected: $expected, got: $names"
        }
    }
}
```

`/dist/build.gradle.kts`:

```kotlin
plugins {
    `java-library`
}

val zips by configurations.creating {
    isCanBeConsumed = true
    isCanBeResolved = false
}

dependencies {
    // This dependency breaks :dist:test task
    implementation(project(":"))
}

// Here we generate a zip file along with its checksum and share it via zips configuration with the root project
val buildZip by tasks.registering(Zip::class) {
    archiveBaseName.set("buildZip")
    archiveVersion.set("")
    from("build.gradle.kts")
}

artifacts {
    add(zips.name, buildZip)
}

val archiveFile = buildZip.flatMap { it.archiveFile }
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }
val shaTask = project.tasks.register(buildZip.name + "Sha512") {
    onlyIf { archiveFile.get().asFile.exists() }
    inputs.file(archiveFile)
    outputs.file(sha512File)
    doLast {
        ant.withGroovyBuilder {
            "checksum"(
                "file" to archiveFile.get(),
                "algorithm" to "SHA-512",
                "fileext" to ".sha512",
                // Make the files verifiable with shasum -c *.sha512
                "format" to "MD5SUM"
            )
        }
    }
}
artifacts {
    // https://github.com/gradle/gradle/issues/10960
    add(zips.name, sha512File) {
        type = "sha512"
        builtBy(shaTask)
    }
}

project.tasks.named(BasePlugin.ASSEMBLE_TASK_NAME) {
    dependsOn(shaTask)
}
```

Run command: `./gradlew :printDistFiles`

```
Caused by: org.gradle.api.InvalidUserCodeException: Querying the mapped value of flatmap(provider(task 'buildZip', class org.gradle.api.tasks.bundling.Zip)) before task ':dist:buildZip' has completed is not supported
	at org.gradle.api.internal.provider.TransformBackedProvider.lambda$beforeRead$0(TransformBackedProvider.java:84)
	at org.gradle.api.internal.provider.ValueSupplier$TaskProducer.visitContentProducerTasks(ValueSupplier.java:136)
	at org.gradle.api.internal.provider.TransformBackedProvider.beforeRead(TransformBackedProvider.java:81)
	at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:63)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:84)
	at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getDelegate(LazyPublishArtifact.java:94)
	at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getName(LazyPublishArtifact.java:64)
	at org.gradle.util.GUtil.getOrDefault(GUtil.java:167)
	at org.gradle.api.internal.artifacts.publish.DecoratingPublishArtifact.getName(DecoratingPublishArtifact.java:51)
	at org.gradle.api.internal.artifacts.publish.DecoratingPublishArtifact_Decorated.getName(Unknown Source)
	at org.gradle.internal.component.model.DefaultIvyArtifactName.forPublishArtifact(DefaultIvyArtifactName.java:36)
	at org.gradle.internal.component.local.model.PublishArtifactLocalArtifactMetadata.<init>(PublishArtifactLocalArtifactMetadata.java:40)
	at org.gradle.internal.component.local.model.DefaultLocalComponentMetadata.addArtifacts(DefaultLocalComponentMetadata.java:147)
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultLocalComponentMetadataBuilder$1.visitArtifacts(DefaultLocalComponentMetadataBuilder.java:56)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfigurationPublications.collectVariants(DefaultConfigurationPublications.java:88)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.collectVariants(DefaultConfiguration.java:1052)
	at org.gradle.api.internal.artifacts.ivyservice.moduleconverter.DefaultLocalComponentMetadataBuilder.addConfiguration(DefaultLocalComponentMetadataBuilder.java:53)
	at org.gradle.api.internal.artifacts.ivyservice.projectmodule.DefaultProjectLocalComponentProvider.getLocalComponentMetadata(DefaultProjectLocalComponentProvider.java:103)
	at org.gradle.api.internal.artifacts.ivyservice.projectmodule.DefaultProjectLocalComponentProvider.lambda$getLocalComponentMetadata$0(DefaultProjectLocalComponentProvider.java:90)
	at org.gradle.api.internal.project.DefaultProjectStateRegistry$ProjectStateImpl.fromMutableState(DefaultProjectStateRegistry.java:267)
	at org.gradle.api.internal.artifacts.ivyservice.projectmodule.DefaultProjectLocalComponentProvider.getLocalComponentMetadata(DefaultProjectLocalComponentProvider.java:87)
	at org.gradle.api.internal.artifacts.ivyservice.projectmodule.DefaultProjectLocalComponentProvider.getComponent(DefaultProjectLocalComponentProvider.java:73)
	at org.gradle.api.internal.artifacts.ivyservice.projectmodule.DefaultLocalComponentRegistry.getComponent(DefaultLocalComponentRegistry.java:34)
	at org.gradle.api.internal.artifacts.DefaultComponentSelectorConverter.getSelector(DefaultComponentSelectorConverter.java:59)
	at org.gradle.api.internal.artifacts.DefaultComponentSelectorConverter.getModule(DefaultComponentSelectorConverter.java:47)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DependencyState.getModuleIdentifier(DependencyState.java:80)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DefaultPendingDependenciesVisitor.maybeAddAsPendingDependency(DefaultPendingDependenciesVisitor.java:35)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.NodeState.visitDependencies(NodeState.java:438)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.NodeState.visitOutgoingDependencies(NodeState.java:284)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DependencyGraphBuilder.traverseGraph(DependencyGraphBuilder.java:194)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder.DependencyGraphBuilder.resolve(DependencyGraphBuilder.java:148)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.DefaultArtifactDependencyResolver.resolve(DefaultArtifactDependencyResolver.java:145)
	at org.gradle.api.internal.artifacts.ivyservice.DefaultConfigurationResolver.resolveBuildDependencies(DefaultConfigurationResolver.java:136)
	at org.gradle.api.internal.artifacts.ivyservice.ShortCircuitEmptyConfigurationResolver.resolveBuildDependencies(ShortCircuitEmptyConfigurationResolver.java:76)
	at org.gradle.api.internal.artifacts.ivyservice.ErrorHandlingConfigurationResolver.resolveBuildDependencies(ErrorHandlingConfigurationResolver.java:63)
```

### Context

The pattern is used in Apache JMeter build.
"core, plugins" projects produce jar files via regular java-library
"dist" project aggregates all the projects and builds distzip
"license" project takes "dist.runtimeElements" and builds a license file for the third-party dependencies
"dist" project includes the license file into the distribution as well.

The thing almost works in Gradle 7, however, lazy artifacts defeat it :(

I am not sure if this issue is significantly different from https://github.com/gradle/gradle/issues/14236, however, now I have a standalone reproducer.

### Steps to Reproduce 

Here's a reproducer: https://github.com/vlsi/gradle-issue-reproducers/pull/5
The structure there is:
`:dist` project produces a `buildZip.zip` into `zips` configuration.
`:dist` project computes SHA512 for `buildZip.zip` and attaches it to `zips` configuration as well.
The root project consumes `:dist:zips` configuration and prints the resulting files.

### Your Environment

Build scan URL for Gradle 6.8.3: https://gradle.com/s/b6cfwbznrq3js
Build scan URL for Gradle 7.0: https://scans.gradle.com/s/4qgffhiqup472

**Comment by vlsi** [comment]
The workaround is to provide all the coordinates when adding an artifact to the configuration like in

```kotlin
artifacts {
    add(zips.name, sha512File) {
        // WA for https://github.com/gradle/gradle/issues/16777
        // The exact values do not seem to be really important, are they?
        name = "testfile"
        classifier = ""
        extension = "sha512"
        // https://github.com/gradle/gradle/issues/10960
        type = "sha512"
        builtBy(shaTask)
    }
}
```

**Comment by bluemonkmn** [comment]
This may be responsible for errors I'm struggling with in trying to set up a new Unreal Quest2 development environment documented at https://stackoverflow.com/questions/68430609/unable-to-set-up-new-unreal-quest-2-development-environment

**Comment by soulskill** [comment]
Gradle `7.1.1` seems to have fixed this issue.

**Comment by anticafe** [comment]
I'm using Gradle Wrapper 7.1.1 and still get that issue.
>Querying the mapped value of task ':ktlintKotlinScriptCheck' property 'reporterOutputDir' before task ':ktlintKotlinScriptCheck' has completed is not supported

Note: I'm also using Ktlint Gradle plugin: `id "org.jlleitschuh.gradle.ktlint" version "9.4.1"`

**Comment by thuytrinh** [comment]
@anticafe [ktlint-gradle v10.1.0](https://github.com/JLLeitschuh/ktlint-gradle/releases/tag/v10.1.0) + Gradle v7.1.1 resolved this issue for me.

**Comment by anticafe** [comment]
Thanks @thuytrinh the combination of **ktlint-gradle v10.1.0 + Gradle v7.1.1** work for me.

**Comment by mahyarmlk** [comment]
I'm using gradle-7.2-rc-3 without ktlint and still get this error.

**Comment by vlsi** [comment]
The issue is still relevant

**Comment by andymalakov** [comment]
Still there 7.4.2

**Comment by tresat** [comment]
tldr; This is the intended behavior and not a bug, but there are various changes you can make to easily fix the issue.

# Overview

```kotlin
val archiveFile = buildZip.flatMap { it.archiveFile }
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }
```
This is *attempting* to express "the output location of `sha512File` is based on the output location of `buildZip.archiveFile`".  We don't actually care about the contents of `buildZip.archiveFile` in this scenario.  

An error is produced because what we've *actually* expressed is a relationship between the files, and also the tasks that produce them (because the mapping preserves task provenance information) - *not just their locations*.  Gradle can't know that we don't care about the file contents here, it only know that it needs an output the `buildZip` task in order to get an output of the `shaTask` task to determine the artifacts present on the `zips` configuration - but the `buildZip` task hasn't yet run.

# Solution 1

If you replace the line `val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }` with: `val sha512File = archiveFile.flatMap { project.provider { File(it.asFile.absolutePath + ".sha512") } }`  Then everything works.  This is because using the `project.provider` factory will create a new `Provider` with no task dependency information and prevent the task information from propagating.  It is generally not a good idea to throw away provenance information like this.

# Solution 2

Interestingly enough, `archiveFile` actually *already returns* a `RegularFilePropery` which contains a `locationOnly` property.  This should allow us to cast the result to that type, and express a relationship on only the location, not the contents of the file. 

```kotlin
val archiveFile = buildZip.flatMap { (it.archiveFile as RegularFileProperty).locationOnly }  
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }
```

The cast succeeds, but running produces a problem with task configuration:

```
* What went wrong:
A problem was found with the configuration of task ':dist:buildZipSha512' (type 'DefaultTask').

    Reason: Task ':dist:buildZipSha512' uses this output of task ':dist:buildZip' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':dist:buildZip' as an input of ':dist:buildZipSha512'.
      2. Declare an explicit dependency on ':dist:buildZip' from ':dist:buildZipSha512' using Task#dependsOn.
      3. Declare an explicit dependency on ':dist:buildZip' from ':dist:buildZipSha512' using Task#mustRunAfter.
```
because we've now broken the task dependency information.  Adding an explicit `dependsOn(buildZip)` to `shaTask` fixes this.  

This relies on a downcast, though, which isn't great.  (Hopefully, [this PR](https://github.com/gradle/gradle/issues/23849) will add a new property to make this explicit and reliable.)

# Solution 3

Replace everything from the `val archiveFile =` line to the end of the `artifacts` block with a custom `Sha512` task:

```kotlin
abstract class Sha512 : DefaultTask() {  
	@get:InputFile  
	abstract val archiveFile: RegularFileProperty  
	  
	@get:OutputFile  
	abstract val checksumFile: RegularFileProperty  
	  
	@TaskAction  
	fun checksum() {  
		ant.withGroovyBuilder {  
			"checksum"(  
				"file" to archiveFile.get(),  
				"algorithm" to "SHA-512",  
				"fileext" to ".sha512",  
				// Make the files verifiable with shasum -c *.sha512  
				"format" to "MD5SUM"  
			)  
		}  
	}  
}  
  
val shaTask = project.tasks.register(buildZip.name + "Sha512", Sha512::class.java) {  
	onlyIf { archiveFile.get().asFile.exists() }  
	archiveFile.convention(buildZip.flatMap { it.archiveFile })  
	checksumFile.fileProvider(archiveFile.locationOnly.map { File(it.asFile.absolutePath + ".sha512") })  
}  
  
artifacts {  
	add(zips.name, shaTask.flatMap { it.checksumFile }) {  
		type = "sha512"  
	}  
}
```

This combines parts of both other solutions into something more idiomatic, as the cost of a new task type.

**Comment by vlsi** [comment]
Thanks for the detailed analysis.

I see you classify the issue as "intended behavior and not a bug", however, it would be extremely nice if Gradle could raise user-facing actionable error in those cases.

Currently, Gradle prints a really strange Gradle-only stacktace, and there's almost no clue on **why** the thing fails.

You notice that I am dealing with "file name only", however, that was probably an oversimplification of the test case.

Imagine that `printDistFiles` not just prints the file names, but it builds the zip file. In my full case, it does build a `zip` archive of the input files. In that case, I **do need full file contents**, not just their names.

Imagine a root project had the following:

```kotlin
// This is instead of printDistFiles
val releaseZip by tasks.registering(Zip::class) {
    archiveBaseName.set("releaseZip")
    from(zips)
    doLast {
        // Just make sure all the files are there
        println("zip: ${archiveFile.get().asFile}")
        val names = zips.map { it.name }.sorted()
        val expected = listOf("buildZip.zip", "buildZip.zip.sha512")
        require(names == expected) {
            "expected: $expected, got: $names"
        }
    }
}
```

----

Just in case, your suggestion with `locationOnly` helps, however, I incline it is a workaround since there's absolutely no mental clue on **why** `locationOnly` is needed there.

Consider your "solution 3":

``` kotlin
val shaTask = project.tasks.register(buildZip.name + "Sha512", Sha512::class.java) {  
	onlyIf { archiveFile.get().asFile.exists() }  
	archiveFile.convention(buildZip.flatMap { it.archiveFile })  
	checksumFile.fileProvider(archiveFile.locationOnly.map { File(it.asFile.absolutePath + ".sha512") })  
}  
```

Here we see at **task usage**. Does it really sound sane for all the users of the "well-defined SHA task" to know of this Gradle quirk? How do they know they must add `localtionOnly` exactly at that place?
Why would they add `locationOnly` at all? The documentation clearly reads

>This should only be used when the task does, in fact, not use the content of this file system element

On contrary, `sha512` task must use the contents of the file system element, otherwise, it won't able to hash it.
So the use of `locationOnly` at that stage would really look strange to the users.
Ok, ok, we talk on the output property only, however, well, do you really consider that an idiomatic and user-friendly engine behaviour?

----

After playing with `locationOnly` a little, and taking into account that the workaround was to specify name, extension, and classifier explicitly within `artifacts.add`, I realized the culprit is that `add(configurationName, file)` uses "file-only" part from the given file.

I incline that Gradle could do a little better by using internal APIs for something like "get filename only from the provider" at lines around

```
	at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getName(LazyPublishArtifact.java:64)
	at org.gradle.util.GUtil.getOrDefault(GUtil.java:167)
	at org.gradle.api.internal.artifacts.publish.DecoratingPublishArtifact.getName(DecoratingPublishArtifact.java:51)
	at org.gradle.api.internal.artifacts.publish.DecoratingPublishArtifact_Decorated.getName(Unknown Source)
	at org.gradle.internal.component.model.DefaultIvyArtifactName.forPublishArtifact(DefaultIvyArtifactName.java:36)
	at org.gradle.internal.component.local.model.PublishArtifactLocalArtifactMetadata.<init>(PublishArtifactLocalArtifactMetadata.java:40)
	at org.gradle.internal.component.local.model.DefaultLocalComponentMetadata.addArtifacts(DefaultLocalComponentMetadata.java:147)
```

In other words, they (e.g. `PublishArtifactLocalArtifactMetadata`) need a name only. Can it just ask the providers and infer "name only"? Why does it care about the task dependencies at that time?

WDYT?

**Comment by vlsi** [comment]
A side note: 

```kotlin
val sha512File = archiveFile.flatMap { project.provider { File(it.asFile.absolutePath + ".sha512") } }
```

This indeed works, however, if I try

```kotlin
val archiveFile = buildZip.flatMap { objects.fileProperty().value(it.archiveFile) }
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }
val sha512FileNameOnly = sha512File.flatMap { provider { it as File } }
```

Then the error is still there:

```
Caused by: org.gradle.api.InvalidUserCodeException: Querying the mapped value of flatmap(provider(task 'buildZip', class org.gradle.api.tasks.bundling.Zip)) before task ':dist:buildZip' has completed is not supported
        at org.gradle.api.internal.provider.TransformBackedProvider.lambda$beforeRead$0(TransformBackedProvider.java:84)
        at org.gradle.api.internal.provider.ValueSupplier$TaskProducer.visitContentProducerTasks(ValueSupplier.java:136)
        at org.gradle.api.internal.provider.TransformBackedProvider.beforeRead(TransformBackedProvider.java:81)
        at org.gradle.api.internal.provider.TransformBackedProvider.calculateOwnValue(TransformBackedProvider.java:63)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
        at org.gradle.api.internal.provider.FlatMapProvider.calculateOwnValue(FlatMapProvider.java:46)
        at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:84)
        at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getDelegate(LazyPublishArtifact.java:93)
        at org.gradle.api.internal.artifacts.dsl.LazyPublishArtifact.getName(LazyPublishArtifact.java:63)
```

So it looks like `.flatMap { provider { ... } }` does not always enables dropping the task dependencies of a provider.
However, that is a slightly different topic.

**Comment by vlsi** [comment]
Here's yet another variation does not allow to drop task dependencies:

```kotlin
val archiveFile = buildZip.flatMap { objects.fileProperty().value(it.archiveFile) }
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }

val sha512FileNameOnly = objects.fileProperty().fileProvider(sha512File).locationOnly
```

**Comment by vlsi** [comment]
It looks like the first `buildZip.flatMap` somehow freezes the dependencies, and there's no way to drop the dependencies after that.

I incline it might be better to add an explicit API for dropping the dependencies, and `.flatMap` should carry all the dependencies by default.
It looks like it would be easier to reason:
a) all `.map` and `.flatMap` carry all the dependencies
b) if you need to drop dependencies, you'll have to add an explicit `.withoutTaskDependencies`

**Comment by big-guy** [comment]
I think making `getArchiveFile` in `AbstractArchiveTask` return just a `Provider<RegularFile>` was a mistake.  We'll do something about that in #23849, but I'm not sure when.

tl;dr:
For now, it's safe to cast `archiveFile` to a `RegularFileProperty`:

That changes:
```
val sha512File = archiveFile.map { File(it.asFile.absolutePath + ".sha512") }
```
to:
```
val sha512File = buildZip.flatMap { (it.archiveFile as RegularFileProperty).locationOnly }.map { File(it.asFile.absolutePath + ".sha512") }
```
This says what is happening: the location of the sha512File is determined by looking at buildZip's archiveFile's location and not its contents. 




It might be nice if `RegularFile` had a succinct way of adding to its basename so you don't need to drop to a File, like:
```
val sha512File = buildZip.flatMap { it.archiveFile.locationOnly }.map { it.append(".sha512")) }
``` 

I don't think we want to introduce a generic "drop dependencies" API to Provider because that conflates together the effect (no dependencies) from the reason (we only care about the location, not the content). IMO, this hides what's really going on and draws attention to the lack of task dependencies:
```
val sha512File = archiveFile.withoutTaskDependencies().map { File(it.asFile.absolutePath + ".sha512") }
```

Providers of file-like things may be somewhat special because there are three aspects to them:
- The build dependencies 
- The content represented by the file
- The location of the file

Most of the time, you only care about the first two things and the location is secondary. 

The behavior of `map` and `flatMap` are different, but they're intended to be:
- `map` keeps the dependencies of the original Provider 
- `flatMap` ignores the dependencies of the original Provider

I think the confusing part of this is that `someTask.flatMap {  }` seems to keep the dependency sometimes. But this is because the Providers returned in the `flatMap` happen to have the same dependencies (`someTask`) and it seems to lose the dependency because sometimes the Provider returned in the `flatMap`, even from a task, do not carry dependency information.

> So it looks like .flatMap { provider { ... } } does not always enables dropping the task dependencies of a provider.

I think this is subtle, but the problem here isn't that `provider { ... }` doesn't "drop dependencies".  It's that we need to calculate the value (the File) to pass to `flatMap` and that triggers the error. 

I'm going to close this as working as-is.  If you have a `FileSystemLocationProperty` and don't need the contents of the files described by the Provider, you can use `getLocationOnly()` to get a Provider for the location and that doesn't carry dependency information for the producer of those files.
